arcsi/api/show.py

```diff
@@ -389,30 +389,52 @@ def view_show_archive(show_slug):
 @auth_token_required
 def view_show_page(show_slug):
     do = DoArchive()
+    archived = request.args.get('archived', True, type=lambda v: v.lower() == 'true')
+    latest = request.args.get('latest', False, type=lambda v: v.lower() == 'true')
     show_query = Show.query.filter_by(archive_lahmastore_base_url=show_slug)
     show = show_query.first()
     if show:
-        # subquery = session.query(Item.id).filter(blabla -timedelta(day=1)).all().subquery()
-        # query = session.query(Show).filter_by(blabla).(Item.id.in_(subquery))
         if show.cover_image_url:
             show.cover_image_url = do.download(
                 show.archive_lahmastore_base_url, show.cover_image_url
             )
         serial_show = show_archive_schema.dump(show)
-        show_items = [
-            show_item
-            for show_item in serial_show["items"]
-            if datetime.strptime(show_item.get("play_date"), "%Y-%m-%d")
-            + timedelta(days=1)
-            < datetime.today()
-        ]
-        serial_show["items"]=show_items
-        for item in serial_show["items"]:
-            item["image_url"] = do.download(
-                show.archive_lahmastore_base_url, item["image_url"]
-            )
-            item["name_slug"]=normalise(item["name"])
-        return serial_show
+        if (archived):
+            show_items = [
+                show_item
+                for show_item in serial_show["items"]
+                if (show_item.get("archived") &
+                    (datetime.strptime(show_item.get("play_date"), "%Y-%m-%d") + timedelta(days=1) < datetime.today()))
+            ]
+            if (latest):
+                show_items = sorted(show_items, key=lambda x: x["play_date"], reverse=True)
+                serial_show["items"] = show_items[0]
+                serial_show["items"]["image_url"] = do.download(show.archive_lahmastore_base_url, serial_show["items"]["image_url"])
+                serial_show["items"]["name_slug"] = normalise(serial_show["items"]["name"])
+                return serial_show
+            else:
+                serial_show["items"] = show_items
```

Comment:
The whole logic of this part is reversed -- it calculates the differences and after that goes through the common parts. 

First we should set up the common parts of `serial_show` then in one step make any adjustments. 

The benefit is that the most common view is handled first and any adjustments can be independently reworked, easily extended or disabled without fears of breaking the default view.

Reply:
I did the refactor for this ugly if-else tree, hope you will like the current version too